### ISSUE TITLE
It takes too much time on DataFrame .loc operation in MLE estimate_cpd method with large states.

### ISSUE THREAD

**Issue by reeingal** [pgmpy/pgmpy#1524]
* pgmpy version: `0.1.18`
* pandas version: `1.4.2`
* Python version: `3.8.8`
* Operating System: `Windows 10`

I make a bayesian network with 25 nodes, but a node have many edges from other nodes. When I use MLE to fit samples, it takes too much time. I debug the code, and find that it gets bogged down in estimate_cpd() in MaximumLikelihoodEstimator for the node with too many edges:

```python
# pgmpy.estimators.MLE.MaximumLikelihoodEstimator.estimate_cpd
state_counts.loc[:, (state_counts == 0).all()] = 1
```

In my case, the `state_counts` is a 2x11943936 `DataFrame`(I known it is a strange struct, but my team leader require use it), and the `.loc` assignment will take too much time (but I don't know why). For now I modify the code to:

```python
import pandas as pd

# ...

tmp_n = state_count.to_numpy()
tmp_n[:, (tmp_n==0).all(axis=0)]=1
state_counts = pd.DataFrame(tmp_n, columns=state_counts.columns)
```

the code above will execute much faster than `state_counts.loc[:, (state_counts == 0).all()] = 1` and the result seems to be same.

**Comment by ankurankan** [comment]
@reeingal Thanks for reporting this. It is indeed weird why there's such a huge difference between comparing values. I did a quick runtime check on the two version and numpy comparison seems to be around 19x faster than pandas.
```python

In [15]: %timeit (df==0).all()
187 ms ± 211 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [16]: %timeit (data==0).all(axis=0)
10.7 ms ± 21.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
I think the best way to fix this would be to replace the line with something like this:
```python
state_counts.loc[:, (state_counts.values==0).all(axis=0)] = 1
```
It's as fast as the modification you suggested and avoids creating extra variables. Would you like to create a PR for this?

**Comment by reeingal** [comment]
> @reeingal Thanks for reporting this. It is indeed weird why there's such a huge difference between comparing values. I did a quick runtime check on the two version and numpy comparison seems to be around 19x faster than pandas.
> 
> ```python
> In [15]: %timeit (df==0).all()
> 187 ms ± 211 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
> 
> In [16]: %timeit (data==0).all(axis=0)
> 10.7 ms ± 21.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
> ```
> 
> I think the best way to fix this would be to replace the line with something like this:
> 
> ```python
> state_counts.loc[:, (state_counts.values==0).all(axis=0)] = 1
> ```
> 
> It's as fast as the modification you suggested and avoids creating extra variables. Would you like to create a PR for this?

It is my pleasure to report a pr.

**Comment by reeingal** [comment]
> @reeingal Thanks for reporting this. It is indeed weird why there's such a huge difference between comparing values. I did a quick runtime check on the two version and numpy comparison seems to be around 19x faster than pandas.
> 
> ```python
> In [15]: %timeit (df==0).all()
> 187 ms ± 211 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
> 
> In [16]: %timeit (data==0).all(axis=0)
> 10.7 ms ± 21.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
> ```
> 
> I think the best way to fix this would be to replace the line with something like this:
> 
> ```python
> state_counts.loc[:, (state_counts.values==0).all(axis=0)] = 1
> ```
> 
> It's as fast as the modification you suggested and avoids creating extra variables. Would you like to create a PR for this?

@ankurankan It is still very slow use `state_counts.loc[:, (state_counts.values==0).all(axis=0)] = 1`. My test code is like:

```python
import logging

import numpy as np
import pandas as pd
from pgmpy.models import BayesianNetwork
from pgmpy.estimators import MaximumLikelihoodEstimator

logging.basicConfig(
    level=logging.INFO,
    format="[%(asctime)s][%(levelname)s] %(message)s"
)

logging.info("Generating random data")

raw_data = np.random.randint(0, 5, (10000, 10))

df = pd.DataFrame(raw_data, columns=[f"feature_{i}" for i in range(1,11)])

edges = [(f"feature_{i}", "feature_10") for i in range(1, 10)] # construct a node with many edges

logging.info("Finish genrate random data ")

model = BayesianNetwork(edges)

logging.info("Begin to train net")

model.fit(
    df,
    estimator=MaximumLikelihoodEstimator,
)
```

I tested the code in my school computer (Windows 10, I9-12900KF, 64GB memory, the pandas version is `1.4.2`) and my laptop (Ubuntu 22.04, AMD 5500U, 16GB, the pandas version is `1.2.4`), and got the same phenomenon. It seems to get bogged in the `loc` assignment of `DataFrame`, maybe a bug of pandas in huge `DataFrame`.

I try to modify the code to `state_counts.values[:, (state_counts.values==0).all(axis=0)] = 1`, it seems work and no extra variables created.

**Comment by ankurankan** [comment]
@reeingal Yes, seem like loc is a bit slower than indexing on the numpy array. The values seem the same to me as well, so should be fine.
